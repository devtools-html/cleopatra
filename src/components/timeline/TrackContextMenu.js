/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import React, { PureComponent } from 'react';
import { ContextMenu, MenuItem } from 'react-contextmenu';
import {
  hideGlobalTrack,
  showGlobalTrack,
  isolateGlobalTrack,
  isolateLocalTrack,
  hideLocalTrack,
  showLocalTrack,
} from '../../actions/profile-view';
import explicitConnect from '../../utils/connect';
import { ensureExists } from '../../utils/flow';
import {
  getThreads,
  getRightClickedTrack,
  getGlobalTracks,
  getRightClickedThreadIndex,
  getLocalTrackNamesByPid,
  getGlobalTrackNames,
} from '../../reducers/profile-view';
import {
  getGlobalTrackOrder,
  getHiddenGlobalTracks,
  getHiddenLocalTracksByPid,
  getLocalTrackOrderByPid,
} from '../../reducers/url-state';
import classNames from 'classnames';

import type { Thread, ThreadIndex, Pid } from '../../types/profile';
import type { TrackIndex, GlobalTrack } from '../../types/profile-derived';
import type { State } from '../../types/reducers';
import type { TrackReference } from '../../types/actions';

import type {
  ExplicitConnectOptions,
  ConnectedProps,
} from '../../utils/connect';

type StateProps = {|
  +threads: Thread[],
  +globalTrackOrder: TrackIndex[],
  +hiddenGlobalTracks: Set<TrackIndex>,
  +hiddenLocalTracksByPid: Map<Pid, Set<TrackIndex>>,
  +localTrackOrderByPid: Map<Pid, TrackIndex[]>,
  +rightClickedTrack: TrackReference,
  +globalTracks: GlobalTrack[],
  +rightClickedThreadIndex: ThreadIndex | null,
  +globalTrackNames: string[],
  +localTrackNamesByPid: Map<Pid, string[]>,
|};

type DispatchProps = {|
  +hideGlobalTrack: typeof hideGlobalTrack,
  +showGlobalTrack: typeof showGlobalTrack,
  +isolateGlobalTrack: typeof isolateGlobalTrack,
  +hideLocalTrack: typeof hideLocalTrack,
  +showLocalTrack: typeof showLocalTrack,
  +isolateLocalTrack: typeof isolateLocalTrack,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class TimelineTrackContextMenu extends PureComponent<Props> {
  _toggleGlobalTrackVisibility = (
    _,
    data: { trackIndex: TrackIndex }
  ): void => {
    const { trackIndex } = data;
    const { hiddenGlobalTracks, hideGlobalTrack, showGlobalTrack } = this.props;
    if (hiddenGlobalTracks.has(trackIndex)) {
      showGlobalTrack(trackIndex);
    } else {
      hideGlobalTrack(trackIndex);
    }
  };

  _toggleLocalTrackVisibility = (
    _,
    data: { pid: Pid, trackIndex: TrackIndex }
  ): void => {
    const { trackIndex, pid } = data;
    const {
      hiddenLocalTracksByPid,
      hideLocalTrack,
      showLocalTrack,
    } = this.props;
    const hiddenLocalTracks = ensureExists(
      hiddenLocalTracksByPid.get(pid),
      'Expected to find hidden local tracks for the given pid'
    );
    if (hiddenLocalTracks.has(trackIndex)) {
      showLocalTrack(pid, trackIndex);
    } else {
      hideLocalTrack(pid, trackIndex);
    }
  };

  _isolateTrack = () => {
    const {
      isolateGlobalTrack,
      isolateLocalTrack,
      rightClickedTrack,
    } = this.props;
    if (rightClickedTrack.type === 'global') {
      isolateGlobalTrack(rightClickedTrack.trackIndex);
    } else {
      const { pid, trackIndex } = rightClickedTrack;
      isolateLocalTrack(pid, trackIndex);
    }
  };

  renderGlobalTrack(trackIndex: TrackIndex) {
    const { hiddenGlobalTracks, globalTrackNames } = this.props;
    const isHidden = hiddenGlobalTracks.has(trackIndex);

    return (
      <MenuItem
        key={trackIndex}
        preventClose={true}
        data={{ trackIndex }}
        onClick={this._toggleGlobalTrackVisibility}
        attributes={{
          className: classNames({ checkable: true, checked: !isHidden }),
        }}
      >
        {globalTrackNames[trackIndex]}
      </MenuItem>
    );
  }

  renderLocalTracks(globalTrackIndex: TrackIndex, pid: Pid) {
    const {
      hiddenLocalTracksByPid,
      localTrackOrderByPid,
      localTrackNamesByPid,
      hiddenGlobalTracks,
    } = this.props;

    const isGlobalTrackHidden = hiddenGlobalTracks.has(globalTrackIndex);
    const localTrackOrder = localTrackOrderByPid.get(pid);
    const hiddenLocalTracks = hiddenLocalTracksByPid.get(pid);
    const localTrackNames = localTrackNamesByPid.get(pid);

    if (
      localTrackOrder === undefined ||
      hiddenLocalTracks === undefined ||
      localTrackNames === undefined
    ) {
      console.error(
        'Unable to find local track information for the given pid:',
        pid
      );
      return null;
    }

    return localTrackOrder.map(trackIndex => (
      <MenuItem
        disabled={isGlobalTrackHidden}
        key={trackIndex}
        preventClose={true}
        data={{ pid, trackIndex }}
        onClick={this._toggleLocalTrackVisibility}
        attributes={{
          className: classNames('checkable indented', {
            checked: !hiddenLocalTracks.has(trackIndex),
          }),
        }}
      >
        {localTrackNames[trackIndex]}
      </MenuItem>
    ));
  }

  getRightClickedTrackName() {
    const {
      globalTrackNames,
      localTrackNamesByPid,
      rightClickedTrack,
    } = this.props;

    if (rightClickedTrack.type === 'global') {
      return globalTrackNames[rightClickedTrack.trackIndex];
    } else {
      const localTrackNames = localTrackNamesByPid.get(rightClickedTrack.pid);
      if (localTrackNames === undefined) {
        console.error('Expected to find a local track name for the given pid.');
        return 'Unknown Track';
      }
      return localTrackNames[rightClickedTrack.trackIndex];
    }
  }

  render() {
    const {
      threads,
      globalTrackOrder,
      hiddenGlobalTracks,
      globalTracks,
      rightClickedThreadIndex,
    } = this.props;

    return (
      <ContextMenu id="TimelineTrackContextMenu">
        {threads.length > 1 && rightClickedThreadIndex !== null ? (
          <div>
            <MenuItem
              onClick={this._isolateTrack}
              disabled={hiddenGlobalTracks.size === globalTrackOrder.length - 1}
            >
              Only show: {`"${this.getRightClickedTrackName()}"`}
            </MenuItem>
            <div className="react-contextmenu-separator" />
          </div>
        ) : null}
        {globalTrackOrder.map(globalTrackIndex => {
          const globalTrack = globalTracks[globalTrackIndex];
          return (
            <div key={globalTrackIndex}>
              {this.renderGlobalTrack(globalTrackIndex)}
              {globalTrack.type === 'process'
                ? this.renderLocalTracks(globalTrackIndex, globalTrack.pid)
                : null}
            </div>
          );
        })}
      </ContextMenu>
    );
  }
}

const options: ExplicitConnectOptions<{||}, StateProps, DispatchProps> = {
  mapStateToProps: (state: State) => ({
    threads: getThreads(state),
    globalTrackOrder: getGlobalTrackOrder(state),
    hiddenGlobalTracks: getHiddenGlobalTracks(state),
    rightClickedTrack: getRightClickedTrack(state),
    globalTracks: getGlobalTracks(state),
    hiddenLocalTracksByPid: getHiddenLocalTracksByPid(state),
    localTrackOrderByPid: getLocalTrackOrderByPid(state),
    rightClickedThreadIndex: getRightClickedThreadIndex(state),
    globalTrackNames: getGlobalTrackNames(state),
    localTrackNamesByPid: getLocalTrackNamesByPid(state),
  }),
  mapDispatchToProps: {
    hideGlobalTrack,
    showGlobalTrack,
    isolateGlobalTrack,
    hideLocalTrack,
    showLocalTrack,
    isolateLocalTrack,
  },
  component: TimelineTrackContextMenu,
};
export default explicitConnect(options);
