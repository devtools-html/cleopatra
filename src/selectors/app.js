/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import { getSelectedTab } from './url-state';
import { tabSlugs } from '../app-logic/tabs-handling';
import { selectedThreadSelectors } from './per-thread';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { AppState, AppViewState } from '../types/state';
import type { Selector } from '../types/store';

/**
 * Simple selectors into the app state.
 */
export const getApp: Selector<AppState> = state => state.app;
export const getView: Selector<AppViewState> = state => getApp(state).view;
export const getIsUrlSetupDone: Selector<boolean> = state =>
  getApp(state).isUrlSetupDone;
export const getHasZoomedViaMousewheel: Selector<boolean> = state => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen: Selector<boolean> = state =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration: Selector<number> = state =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug: Selector<TabSlug> = state =>
  getApp(state).lastVisibleThreadTabSlug;

/**
 * Visible tabs are computed based on the current state of the profile. Some
 * effort is made to not show a tab when there is no data available for it.
 */
export const getVisibleTabs: Selector<$ReadOnlyArray<TabSlug>> = createSelector(
  selectedThreadSelectors.getIsNetworkChartEmptyInFullRange,
  selectedThreadSelectors.getJsTracerTable,
  (isNetworkChartEmpty, jsTracerTable) => {
    let visibleTabs = tabSlugs;
    if (isNetworkChartEmpty) {
      visibleTabs = visibleTabs.filter(tabSlug => tabSlug !== 'network-chart');
    }
    if (!jsTracerTable) {
      visibleTabs = visibleTabs.filter(tabSlug => tabSlug !== 'js-tracer');
    }
    return visibleTabs;
  }
);
