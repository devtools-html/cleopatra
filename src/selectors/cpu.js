/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { createSelector } from 'reselect';

import {
  getThreads,
  getProfileInterval,
  getSampleUnits,
  getMeta,
} from './profile';
import { getThreadSelectors } from './per-thread';
import { computeMaxThreadCPUDelta } from 'firefox-profiler/profile-logic/cpu';

import type { Selector, State, Thread } from 'firefox-profiler/types';

export const getIsCPUUtilizationProvided: Selector<boolean> = createSelector(
  getSampleUnits,
  getMeta,
  getCPUProcessedThreads,
  (sampleUnits, meta, threads) => {
    return (
      sampleUnits !== undefined &&
      // Currently checking the features array for 'cpu' feature should be enough,
      // but in the future we may remove that feature and enable it permanently.
      // Therefore we are also checking the samples table to see if we have CPU
      // delta values.
      ((meta.configuration && meta.configuration.features.includes('cpu')) ||
        threads.some((thread) => thread.samples.threadCPUDelta !== undefined))
    );
  }
);

/**
 * This function returns the list of all threads after the CPU values have been
 * processed. This uses a selector from the per-thread selectors. Because we'll
 * use this selector for every thread, and also need the full state for this call,
 * we can't use the simple memoization from `createSelector`, and instead we
 * need to implement our own simple memoization.
 */
let _threads = null;
let _cpuProcessedThreads = null;
function getCPUProcessedThreads(state: State): Thread[] {
  const threads = getThreads(state);

  if (_threads !== threads || _cpuProcessedThreads === null) {
    // Storing the threads makes it possible to invalidate the memoized value at
    // the right moment.
    _threads = threads;
    _cpuProcessedThreads = threads.map((thread, threadIndex) =>
      getThreadSelectors(threadIndex).getCPUProcessedThread(state)
    );
  }
  return _cpuProcessedThreads;
}

export const getMaxThreadCPUDelta: Selector<number> = createSelector(
  getCPUProcessedThreads,
  getProfileInterval,
  computeMaxThreadCPUDelta
);
