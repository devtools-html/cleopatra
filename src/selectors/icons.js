/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import type { IconWithClassName, IconState } from '../types/state';
import type { Selector, DangerousSelectorWithArguments } from '../types/store';
import type { CallNodeDisplayData } from '../types/profile-derived';

/**
 * A simple selector into the icon state.
 */
export const getIcons: Selector<IconState> = state => state.icons;

/**
 * In order to load icons without multiple requests, icons are created through
 * CSS. This function gets the CSS class name for a call node. This function
 * does not perform any memoization, and updates every time. It could be updated
 * to memoize.
 */
export const getIconClassNameForCallNode: DangerousSelectorWithArguments<
  string,
  CallNodeDisplayData
> = (state, displayData) => {
  const icons = getIcons(state);
  return displayData.icon !== null && icons.has(displayData.icon)
    ? _classNameFromUrl(displayData.icon)
    : '';
};

/**
 * This functions returns an object with both the icon URL and the class name.
 */
export const getIconsWithClassNames: Selector<
  IconWithClassName[]
> = createSelector(
  getIcons,
  icons =>
    [...icons].map(icon => ({ icon, className: _classNameFromUrl(icon) }))
);

/**
 * Transforms a URL into a valid CSS class name.
 */
function _classNameFromUrl(url): string {
  return url.replace(/[/:.+>< ~()#,]/g, '_');
}
