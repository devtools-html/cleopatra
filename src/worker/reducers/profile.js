export default function profile(state = null, action) {
  switch (action.type) {
    case 'PROFILE_PROCESSED':
      return action.profile;
    default:
      return state;
  }
}
