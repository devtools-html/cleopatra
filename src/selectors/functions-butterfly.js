/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { createSelector } from 'reselect';
import { getDefaultCategory } from './profile';
import { selectedThreadSelectors } from './per-thread';
import { getCallNodeInfo } from '../profile-logic/profile-data';

import type {
  IndexIntoFuncTable,
  Thread,
  IndexIntoCategoryList,
} from '../types/profile';
import type { CallNodeInfo } from '../types/profile-derived';
import type { State } from '../types/state';

const EMPTY_ARRAY = [];

const getImplementationFilteredCallNodeInfo = createSelector(
  selectedThreadSelectors.getImplementationFilteredThread,
  getDefaultCategory,
  (
    { stackTable, frameTable, funcTable }: Thread,
    defaultCategory: IndexIntoCategoryList
  ): CallNodeInfo => {
    return getCallNodeInfo(stackTable, frameTable, funcTable, defaultCategory);
  }
);

const getCallingInformationForFunction = createSelector(
  selectedThreadSelectors.getImplementationFilteredThread,
  getImplementationFilteredCallNodeInfo,
  (_, funcIndex: IndexIntoFuncTable | null) => funcIndex,
  (
    thread,
    { callNodeTable },
    funcIndex
  ): {| callers: IndexIntoFuncTable[], callees: IndexIntoFuncTable[] |} => {
    if (funcIndex === null) {
      return { callers: EMPTY_ARRAY, callees: EMPTY_ARRAY };
    }

    const callers = new Set();
    const callees = new Set();

    for (let i = 0; i < callNodeTable.length; i++) {
      const thisFuncIndex = callNodeTable.func[i];
      const callerNodeIndex = callNodeTable.prefix[i];

      if (callerNodeIndex >= 0) {
        const callerFuncIndex = callNodeTable.func[callerNodeIndex];

        if (thisFuncIndex === funcIndex) {
          callers.add(callerFuncIndex);
        }

        if (callerFuncIndex === funcIndex) {
          callees.add(thisFuncIndex);
        }
      }
    }

    return { callers: [...callers], callees: [...callees] };
  }
);

export const getCallersForFunction = (
  state: State,
  funcIndex: IndexIntoFuncTable | null
) => getCallingInformationForFunction(state, funcIndex).callers;
export const getCalleesForFunction = (
  state: State,
  funcIndex: IndexIntoFuncTable | null
) => getCallingInformationForFunction(state, funcIndex).callees;
