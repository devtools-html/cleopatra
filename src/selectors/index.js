/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';

// Normally we don't need that type because flow can infer types, but on that
// object, it gives weird flow errors that say missing type annotation for X.
// Only adding a type like that seems to solve it.
type Selectors = {
  app: typeof app,
  profile: typeof profile,
  urlState: typeof urlState,
  icons: typeof icons,
  publish: typeof publish,
  zippedProfiles: typeof zippedProfiles,
  selectedThread: typeof selectedThread,
};

export default ({
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
}: Selectors);
