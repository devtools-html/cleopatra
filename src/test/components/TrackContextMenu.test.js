/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import {
  changeSelectedThread,
  changeRightClickedTrack,
} from '../../actions/profile-view';
import TrackContextMenu from '../../components/timeline/TrackContextMenu';
import { getGlobalTracks, getLocalTracks } from '../../selectors/profile';
import {
  getHiddenGlobalTracks,
  getHiddenLocalTracks,
} from '../../selectors/url-state';
import {
  getProfileWithNiceTracks,
  getHumanReadableTracks,
} from '../fixtures/profiles/tracks';
import { storeWithProfile } from '../fixtures/stores';

describe('timeline/TrackContextMenu', function() {
  /**
   *  getProfileWithNiceTracks() looks like: [
   *    'show [thread GeckoMain process]',
   *    'show [thread GeckoMain tab]',       <- use this global track.
   *    '  - show [thread DOM Worker]',
   *    '  - show [thread Style]',
   *  ]
   */
  function setup(profile = getProfileWithNiceTracks()) {
    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;

    const view = mount(
      <Provider store={store}>
        <TrackContextMenu />
      </Provider>
    );

    return {
      dispatch,
      getState,
      profile,
      store,
      view,
    };
  }

  describe('selected global track', function() {
    function setupGlobalTrack(profile) {
      const results = setup(profile);
      const { view, dispatch, getState } = results;

      const trackIndex = 1;
      const trackReference = {
        type: 'global',
        trackIndex: trackIndex,
      };
      const track = getGlobalTracks(getState())[trackIndex];
      if (track.type !== 'process') {
        throw new Error('Expected a process track.');
      }
      const threadIndex = track.mainThreadIndex;
      if (threadIndex !== null) {
        // Explicitly select the global thread. Tests can pass in a custom profile,
        // so don't fail if this doesn't exist.
        dispatch(changeSelectedThread(threadIndex));
      }
      dispatch(changeRightClickedTrack(trackReference));
      view.update();

      const isolateProcessItem = view.find(
        '[data-test-id="isolate-track-process"]'
      );
      const isolateProcessMainThreadItem = view.find(
        '[data-test-id="isolate-process-main-thread"]'
      );
      const trackItem = view.find(
        `[data-test-id="global-track-${trackIndex}"]`
      );

      return {
        ...results,
        trackReference,
        trackIndex,
        threadIndex,
        isolateProcessItem,
        isolateProcessMainThreadItem,
        trackItem,
      };
    }

    it('matches the snapshot of a global track', () => {
      const { view } = setupGlobalTrack();
      expect(view).toMatchSnapshot();
    });

    it('has the correct selectors into useful parts of the component', function() {
      const {
        isolateProcessMainThreadItem,
        isolateProcessItem,
        trackItem,
        getState,
      } = setupGlobalTrack();
      expect(isolateProcessMainThreadItem.text()).toBe(
        'Only show "Content Process"'
      );
      expect(isolateProcessItem.text()).toBe('Only show this process');
      expect(trackItem.text()).toBe('Content Process');
      expect(getHumanReadableTracks(getState())).toEqual([
        'show [thread GeckoMain process]',
        'show [thread GeckoMain tab] SELECTED',
        '  - show [thread DOM Worker]',
        '  - show [thread Style]',
      ]);
    });

    it('can isolate the process', function() {
      const { isolateProcessItem, getState } = setupGlobalTrack();
      isolateProcessItem.simulate('click');
      expect(getHumanReadableTracks(getState())).toEqual([
        'hide [thread GeckoMain process]',
        'show [thread GeckoMain tab] SELECTED',
        '  - show [thread DOM Worker]',
        '  - show [thread Style]',
      ]);
    });

    it("can isolate the process's main thread", function() {
      const { isolateProcessMainThreadItem, getState } = setupGlobalTrack();
      isolateProcessMainThreadItem.simulate('click');
      expect(getHumanReadableTracks(getState())).toEqual([
        'hide [thread GeckoMain process]',
        'show [thread GeckoMain tab] SELECTED',
        '  - hide [thread DOM Worker]',
        '  - hide [thread Style]',
      ]);
    });

    it('isolates a process track without a main thread', function() {
      const profile = getProfileWithNiceTracks();
      // Remove the thread [thread GeckoMain tab]
      profile.threads.splice(1, 1);
      const {
        isolateProcessMainThreadItem,
        isolateProcessItem,
        getState,
      } = setupGlobalTrack(profile);

      expect(getHumanReadableTracks(getState())).toEqual([
        'show [thread GeckoMain process] SELECTED',
        'show [process]',
        '  - show [thread DOM Worker]',
        '  - show [thread Style]',
      ]);

      expect(isolateProcessMainThreadItem.length).toBe(0);
      expect(isolateProcessItem.length).toBe(1);
      isolateProcessItem.simulate('click');
      expect(getHumanReadableTracks(getState())).toEqual([
        'hide [thread GeckoMain process]',
        'show [process]',
        '  - show [thread DOM Worker] SELECTED',
        '  - show [thread Style]',
      ]);
    });

    it('can toggle a global track by clicking it', function() {
      const { trackItem, trackIndex, getState } = setupGlobalTrack();
      expect(getHiddenGlobalTracks(getState()).has(trackIndex)).toBe(false);
      trackItem.simulate('click');
      expect(getHiddenGlobalTracks(getState()).has(trackIndex)).toBe(true);
      trackItem.simulate('click');
      expect(getHiddenGlobalTracks(getState()).has(trackIndex)).toBe(false);
    });

    xit('can present a disabled isolate item on non-process tracks', function() {
      // TODO - We should wait until we have some real tracks without a thread index.
    });
  });

  describe('selected local track', function() {
    function setupLocalTrack() {
      const results = setup();
      const { view, dispatch, getState } = results;

      // In getProfileWithNiceTracks, the two pids are 111 and 222 for the
      // "GeckoMain process" and "GeckoMain tab" respectively. Use 222 since it has
      // local tracks.
      const pid = 222;
      const trackIndex = 0;
      const trackReference = {
        type: 'local',
        pid,
        trackIndex,
      };
      const localTracks = getLocalTracks(getState(), pid);
      const localTrack = localTracks[trackIndex];
      if (localTrack.type !== 'thread') {
        throw new Error('Expected a thread track');
      }
      const threadIndex = localTrack.threadIndex;

      // Explicitly select the global thread.
      dispatch(changeSelectedThread(threadIndex));
      dispatch(changeRightClickedTrack(trackReference));
      view.update();

      const isolateLocalTrackItem = view.find(
        '[data-test-id="isolate-local-track"]'
      );
      const trackItem = view.find(
        `[data-test-id="local-track-${pid}-${trackIndex}"]`
      );

      return {
        ...results,
        trackReference,
        trackIndex,
        threadIndex,
        isolateLocalTrackItem,
        trackItem,
        pid,
      };
    }

    it('matches the snapshot of a local track', () => {
      const { view } = setupLocalTrack();
      expect(view).toMatchSnapshot();
    });

    it('has the correct selectors into useful parts of the component', function() {
      const { isolateLocalTrackItem, trackItem, getState } = setupLocalTrack();
      expect(isolateLocalTrackItem.text()).toBe('Only show "DOM Worker"');
      expect(trackItem.text()).toBe('DOM Worker');
      expect(getHumanReadableTracks(getState())).toEqual([
        'show [thread GeckoMain process]',
        'show [thread GeckoMain tab]',
        '  - show [thread DOM Worker] SELECTED',
        '  - show [thread Style]',
      ]);
    });

    it('can isolate the local track', function() {
      const { isolateLocalTrackItem, getState } = setupLocalTrack();
      isolateLocalTrackItem.simulate('click');
      expect(getHumanReadableTracks(getState())).toEqual([
        'hide [thread GeckoMain process]',
        'show [thread GeckoMain tab]',
        '  - show [thread DOM Worker] SELECTED',
        '  - hide [thread Style]',
      ]);
    });

    it('can toggle a local track by clicking it', function() {
      const { trackItem, pid, trackIndex, getState } = setupLocalTrack();
      expect(getHiddenLocalTracks(getState(), pid).has(trackIndex)).toBe(false);
      trackItem.simulate('click');
      expect(getHiddenLocalTracks(getState(), pid).has(trackIndex)).toBe(true);
      trackItem.simulate('click');
      expect(getHiddenLocalTracks(getState(), pid).has(trackIndex)).toBe(false);
    });

    xit('can isolate a non-thread track, as long as there process has a thread index', function() {
      // TODO - We should wait until we have some real non-thread tracks
    });
  });
});
