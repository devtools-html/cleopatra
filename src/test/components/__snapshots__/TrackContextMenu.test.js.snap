// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/TrackContextMenu selected global track matches the snapshot of a global track 1`] = `
<nav
  className="react-contextmenu timeline-context-menu"
  onContextMenu={[Function]}
  onMouseLeave={[Function]}
  role="menu"
  style={
    Object {
      "opacity": 0,
      "pointerEvents": "none",
      "position": "fixed",
    }
  }
  tabIndex="-1"
>
  <div
    aria-disabled="false"
    aria-orientation={null}
    className="react-contextmenu-item"
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    Only show 
    "Content Process"
  </div>
  <div
    aria-disabled="false"
    aria-orientation={null}
    className="react-contextmenu-item"
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    Only show this process
  </div>
  <div
    className="react-contextmenu-separator"
    key=".3"
  />
  <div
    key=".4:$0"
  >
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable checked"
      data-test-id="global-track-0"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      GeckoMain
    </div>
  </div>
  <div
    key=".4:$1"
  >
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable checked"
      data-test-id="global-track-1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      Content Process
    </div>
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable indented checked"
      data-test-id="local-track-222-0"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      DOM Worker
    </div>
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable indented checked"
      data-test-id="local-track-222-1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      Style
    </div>
  </div>
</nav>
`;

exports[`timeline/TrackContextMenu selected local track matches the snapshot of a local track 1`] = `
<nav
  className="react-contextmenu timeline-context-menu"
  onContextMenu={[Function]}
  onMouseLeave={[Function]}
  role="menu"
  style={
    Object {
      "opacity": 0,
      "pointerEvents": "none",
      "position": "fixed",
    }
  }
  tabIndex="-1"
>
  <div
    aria-disabled="false"
    aria-orientation={null}
    className="react-contextmenu-item"
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    role="menuitem"
    tabIndex="-1"
  >
    Only show 
    "DOM Worker"
  </div>
  <div
    className="react-contextmenu-separator"
    key=".3"
  />
  <div
    key=".4:$0"
  >
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable checked"
      data-test-id="global-track-0"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      GeckoMain
    </div>
  </div>
  <div
    key=".4:$1"
  >
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable checked"
      data-test-id="global-track-1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      Content Process
    </div>
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable indented checked"
      data-test-id="local-track-222-0"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      DOM Worker
    </div>
    <div
      aria-disabled="false"
      aria-orientation={null}
      className="react-contextmenu-item checkable indented checked"
      data-test-id="local-track-222-1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onTouchEnd={[Function]}
      role="menuitem"
      tabIndex="-1"
    >
      Style
    </div>
  </div>
</nav>
`;
